### PR TITLE
Refactor the gittag plugin + fix sources of gittag

### LINKED ISSUE #467: Error `no Git tag found matching pattern \"latest\"` with gitTag resources
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

- If you run the current example at `examples/updatecli.generic/gitTag.yaml`, it only errors for every gittag resource
- Example pre-0.17.0 (from documentation) does not work either
- When updating to post-0.17.0 scm syntax, the error keep happening

### Expected Behavior

_No response_

### Steps To Reproduce

- Pre 0.17.0 (from website):

```yaml
title: "Test new gitTag resource"
sources:
  git:
    name: Get Latest updatecli test
    kind: gitTag
    spec:
      url: "https://github.com/updatecli/updatecli.git"
      branch: "main"
```

```shell
$ updatecli diff --config examples/updatecli.generic/gitTag.yaml

+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "examples/updatecli.generic/gitTag.yaml"
Repository retrieved: 0

++++++++++++
+ PIPELINE +
++++++++++++

############################
# TEST NEW GITTAG RESOURCE #
############################

SOURCES
=======
git
---
WARNING: no git tag message specified
Searching for version matching pattern "latest"
✗ No Git Tag found matching pattern "latest"
ERROR: ✗ no Git tag found matching pattern "latest"
Pipeline "Test new gitTag resource" failed
Skipping due to:
        "sources stage:\t\"no Git tag found matching pattern \\\"latest\\\"\""
sources stage:  "no Git tag found matching pattern \"latest\""

=============================
REPORTS:

✗ GITTAG.YAML:
        Sources:
                ✗ [git] Get Latest updatecli test(gitTag)
        Target:

Run Summary
===========
Pipeline(s) run:
  * Changed:    0
  * Failed:     1
  * Skipped:    0
  * Succeeded:  0
  * Total:      1
ERROR: ✗ 1 over 1 pipeline failed
ERROR: command failed
```

* Same with "updated" syntax:

```yaml
title: "Test new gitTag resource"

scms:
  updatecli:
    kind: git
    spec:
      url: "https://github.com/updatecli/updatecli.git"
      branch: "main"

sources:
  git:
    name: Get Latest updatecli test
    kind: gitTag
    scmID: updatecli
```

```shell
updatecli diff --config examples/updatecli.generic/gitTag.yaml 

+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "examples/updatecli.generic/gitTag.yaml"
Repository retrieved: 1


++++++++++++
+ PIPELINE +
++++++++++++



############################
# TEST NEW GITTAG RESOURCE #
############################


SOURCES
=======

git
---
WARNING: no git tag message specified
Searching for version matching pattern "latest"
✗ No Git Tag found matching pattern "latest"
ERROR: ✗ no Git tag found matching pattern "latest"
Pipeline "Test new gitTag resource" failed
Skipping due to:
        "sources stage:\t\"no Git tag found matching pattern \\\"latest\\\"\""
sources stage:  "no Git tag found matching pattern \"latest\""

=============================

REPORTS:


✗ GITTAG.YAML:
        Sources:
                ✗ [git] Get Latest updatecli test(gitTag)
        Target:



Run Summary
===========
Pipeline(s) run:
  * Changed:    0
  * Failed:     1
  * Skipped:    0
  * Succeeded:  0
  * Total:      1
ERROR: ✗ 1 over 1 pipeline failed
ERROR: command failed
```

### Environment

```markdown
- OS: macOS, Linux
- updatecli: 0.18.2
```

### Anything else?

This is only reporting an issue. I intend to fix this as part of https://github.com/updatecli/updatecli/pull/457 unless you disagree.

### LINKED ISSUE #469: The gitTag resource does returns the correct version
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

The gittag does not return the latest tag, but another.
It seems related to a sorting issue (the returned tag are always the "latest" on on lexicographic order).

### Expected Behavior

Example with the updatecli GitHub repository: it returns `v0.9.0` while `v0.18.2` is expected (when writing these lines).

### Steps To Reproduce

```yaml
title: "Test new gitTag resource"

scms:
  updatecli:
    kind: git
    spec:
      url: "https://github.com/updatecli/updatecli.git"
      branch: "main"

sources:
  git:
    name: Get Latest updatecli test
    kind: gitTag
    scmID: updatecli
```

### Environment

```markdown
- OS: macOS, Linux
- updatecli: 0.18.2
```

### Anything else?

This is only reporting an issue. I intend to fix this as part of #457 unless you disagree.

### PR DESCRIPTION
**Pull request by dduportal**
This PR comes from the work in progress in #412 and refactors the gittag plugin into its own package.

While working on this, 2 bugs where found on this plugin and reported (#467 #468). These 2 bugs are now corrected 

Fix #467
Fix #469 

## Test

To test this pull request, you can run the following commands:

```shell
go build -o ./dist/updatecli && ./dist/updatecli diff --config examples/updatecli.generic/gitTag.yaml
```

should error only on the 2 targets (expected) with:

```
ERROR: Git push tag error: authentication required
ERROR: Something went wrong in target "git" :
ERROR: authentication required
```

## Additional Information

- Please note that I did work on the conditions/targets, neither on unit tests for this one
  - ~factorizing code on condition and target between "normal" and "SCM" methods~
  - Using unit test with mock
- The fix embeded on this PR would allow the scenario "get latest stable nginx" (it wasn't possible before):

```yaml
title: Bump nginx:stable docker image version

scms:
  default:
    kind: github
    spec:
      user: "{{ .github.user }}"
      email: "{{ .github.email }}"
      owner: "{{ .github.owner }}"
      repository: "{{ .github.repository }}"
      token: "{{ requiredEnv .github.token }}"
      username: "{{ .github.username }}"
      branch: "{{ .github.branch }}"
  nginx-gh-mirror:
    kind: git
    spec:
      url: "https://github.com/nginx/nginx.git"
      branch: "master"

sources:
  latestRelease:
    name: Get latest stable version of nginx
    kind: gitTag
    scmID: nginx-gh-mirror
    spec:
      versionFilter:
        kind: regex
        ## Nginx stable version have the minor digit as an even number
        pattern: 'release-(\d+)\.(\d*[0|2|4|6|8])\.(\d+)'
    transformers:
      - trimPrefix: "release-"
      - addSuffix: "-alpine"
```

### CODE REVIEW

**Comment by dduportal** [comment]
Special addon for you @olblak : after merging https://github.com/updatecli/updatecli/pull/419, I've added a e2e test for gittag sources :)

**Comment by dduportal** [comment]
OK, I lied: adding a few more changes to factorize the condition/target logic (with or without SCM).

Cannot add unit tests: added a discussion for that

**Review by olblak — DISMISSED** [review]
This is really nice work, thanks @dduportal

File: pkg/plugins/gittag/condition.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+			gt.spec.Path,
+			path)
+	}
+	gt.spec.Path = path
+
+	return gt.condition(source)
+}
+
+func (gt *GitTag) condition(source string) (bool, error) {
+	// If source input is empty, then it as disabled by the user with `disablesourceinput: true`
```
Comment (lemeurherve) [inline pkg/plugins/gittag/condition.go:32]: There is a problem with "it as disabled"

File: pkg/plugins/gittag/condition.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/updatecli/updatecli/pkg/core/result"
+	"github.com/updatecli/updatecli/pkg/plugins/git/generic"
+)
+
+// Condition checks that a git tag exists
+func (gt *GitTag) Condition(source string) (bool, error) {
+	return gt.condition(source)
+}
+
+// ConditionFromSCM test if a tag exist from a git repository specific from SCM
```
Comment (lemeurherve) [inline pkg/plugins/gittag/condition.go:17]:
```suggestion
// ConditionFromSCM test if a tag exists from a git repository specific from SCM
```

File: pkg/plugins/gittag/condition.go (hunk 1 above)
Comment (olblak) [reply pkg/plugins/gittag/condition.go:32]:
```suggestion
	// If source input is empty, then it means that it was disabled by the user with `disablesourceinput: true`
```

**Review by olblak — DISMISSED** [review]

File: e2e/updatecli.d/source/gitTag.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      url: "https://github.com/updatecli/updatecli.git"
+      branch: "main"
+
+sources:
+  git:
+    name: Get Latest updatecli test
+    kind: gitTag
+    scmID: updatecli
+  githubRegex:
+    name: Get Latest updatecli test
```
Comment (lemeurherve) [inline e2e/updatecli.d/source/gitTag.yaml:16]:
```suggestion
    name: Get latest updatecli test
    kind: gitTag
    scmID: updatecli
  githubRegex:
    name: Get latest updatecli test
```

File: pkg/plugins/gittag/condition.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	gt.spec.Path = path
+
+	return gt.condition(source)
+}
+
+func (gt *GitTag) condition(source string) (bool, error) {
+	// If source input is empty, then it means that it was disabled by the user with `disablesourceinput: true`
+	if source != "" {
+		logrus.Infof("Source Input Value detected: using it as spec.versionfilter.pattern")
```
Comment (lemeurherve) [inline pkg/plugins/gittag/condition.go:34]:
```suggestion
		logrus.Infof("Source input value detected: using it as spec.versionfilter.pattern")
```

File: pkg/plugins/gittag/condition.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		err = fmt.Errorf("No Git Tag matching pattern %q, found", gt.spec.VersionFilter.Pattern)
+		return false, err
+	}
+
+	if tag == gt.spec.VersionFilter.Pattern {
+		logrus.Printf("%s Git Tag %q matching\n", result.SUCCESS, gt.spec.VersionFilter.Pattern)
+		return true, nil
+	}
+
+	logrus.Printf("%s Git Tag %q not matching %q\n",
```
Comment (lemeurherve) [inline pkg/plugins/gittag/condition.go:64]:
```suggestion
		err = fmt.Errorf("No git tag matching pattern %q, found", gt.spec.VersionFilter.Pattern)
		return false, err
	}

	if tag == gt.spec.VersionFilter.Pattern {
		logrus.Printf("%s Git tag %q matching\n", result.SUCCESS, gt.spec.VersionFilter.Pattern)
		return true, nil
	}

	logrus.Printf("%s Git Tag %q not matching %q\n",
```

File: pkg/plugins/gittag/main.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+	"github.com/updatecli/updatecli/pkg/plugins/version"
+)
+
+// Spec defines a specification for a "gittag" resource
+// parsed from an updatecli manifest file
+type Spec struct {
+	Path          string         // Path contains the git repository path
+	VersionFilter version.Filter // VersionFilter provides parameters to specify version pattern and its type like regex, semver, or just latest.
+	Message       string         // Message associated to the git Tag
```
Comment (lemeurherve) [inline pkg/plugins/gittag/main.go:15]:
```suggestion
	Message       string         // Message associated to the git tag
```

File: pkg/plugins/gittag/main.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	return &GitTag{
+		spec: spec,
+	}, nil
+}
+
+// Validate tests that tag struct is correctly configured
+func (gt *GitTag) Validate() error {
+	validationErrors := []string{}
+	if gt.spec.Path == "" {
+		validationErrors = append(validationErrors, "Git working directory path is empty while it must be specified. Did you specify an `scmID` or a `spec.path`?")
```
Comment (lemeurherve) [inline pkg/plugins/gittag/main.go:36]:
```suggestion
		validationErrors = append(validationErrors, "Git working directory path is empty while it must be specified. Did you specified an `scmID` or a `spec.path`?")
```

File: pkg/plugins/gittag/main.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	err := gt.spec.VersionFilter.Validate()
+	if err != nil {
+		validationErrors = append(validationErrors, err.Error())
+	}
+
+	// Return all the validation errors if found any
+	if len(validationErrors) > 0 {
+		return fmt.Errorf("Validation error: the provided manifest configuration had the following validation errors:\n%s", strings.Join(validationErrors, "\n\n"))
```
Comment (lemeurherve) [inline pkg/plugins/gittag/main.go:46]:
```suggestion
		return fmt.Errorf("Validation error: the provided manifest configuration has the following validation errors:\n%s", strings.Join(validationErrors, "\n\n"))
```

File: pkg/plugins/gittag/source.go [hunk 9]
```diff
@@ -9,36 +9,34 @@ import (
 )
 
 // Source return the latest git tag based on create time
```
Comment (lemeurherve) [inline pkg/plugins/gittag/source.go:11]:
```suggestion
// Source returns the latest git tag based on create time
```

File: pkg/plugins/gittag/source.go [hunk 10]
```diff
@@ -9,36 +9,34 @@ import (
...
 		return "", err
 	}
-	value := t.foundVersion.ParsedVersion
+	value := gt.foundVersion.ParsedVersion
 
 	if len(value) == 0 {
-		logrus.Infof("%s No Git Tag found matching pattern %q", result.FAILURE, t.VersionFilter.Pattern)
-		return value, fmt.Errorf("no Git tag found matching pattern %q", t.VersionFilter.Pattern)
+		logrus.Infof("%s No Git Tag found matching pattern %q", result.FAILURE, gt.spec.VersionFilter.Pattern)
+		return value, fmt.Errorf("no Git tag found matching pattern %q", gt.spec.VersionFilter.Pattern)
```
Comment (lemeurherve) [inline pkg/plugins/gittag/source.go:37]:
```suggestion
		logrus.Infof("%s No git tag found matching pattern %q", result.FAILURE, gt.spec.VersionFilter.Pattern)
		return value, fmt.Errorf("No git tag found matching pattern %q", gt.spec.VersionFilter.Pattern)
```

File: pkg/plugins/gittag/source.go [hunk 11]
```diff
@@ -9,36 +9,34 @@ import (
...
 	} else if len(value) > 0 {
-		logrus.Infof("%s Git Tag %q found, matching pattern %q", result.SUCCESS, value, t.VersionFilter.Pattern)
+		logrus.Infof("%s Git Tag %q found, matching pattern %q", result.SUCCESS, value, gt.spec.VersionFilter.Pattern)
```
Comment (lemeurherve) [inline pkg/plugins/gittag/source.go:39]:
```suggestion
		logrus.Infof("%s Git tag %q found matching pattern %q", result.SUCCESS, value, gt.spec.VersionFilter.Pattern)
```

File: pkg/plugins/gittag/target.go [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+// Target create a tag if needed from a local git repository, without pushing the tag
+func (gt *GitTag) Target(source string, dryRun bool) (changed bool, err error) {
+	changed, _, _, err = gt.target(source, dryRun)
+
+	return changed, err
+}
+
+// TargetFromSCM create and push a git tag based on the SCM configuration
```
Comment (lemeurherve) [inline pkg/plugins/gittag/target.go:19]:
```suggestion
// Target creates a tag if needed from a local git repository, without pushing the tag
func (gt *GitTag) Target(source string, dryRun bool) (changed bool, err error) {
	changed, _, _, err = gt.target(source, dryRun)

	return changed, err
}

// TargetFromSCM creates and pushes a git tag based on the SCM configuration
```

File: pkg/plugins/gittag/target.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	if err != nil {
+		return changed, files, message, err
+	}
+
+	err = scm.PushTag(source)
+	if err != nil {
+		logrus.Errorf("Git push tag error: %s", err)
+		return changed, files, message, err
+	}
+	logrus.Infof("%s The Git Tag %q was pushed successfully to the specified remote.", result.ATTENTION, source)
```
Comment (lemeurherve) [inline pkg/plugins/gittag/target.go:38]:
```suggestion
	logrus.Infof("%s The git tag %q was pushed successfully to the specified remote.", result.ATTENTION, source)
```

File: pkg/plugins/gittag/target.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+	files := []string{}
+	message := gt.spec.Message
+
+	// Fail if a pattern is specified
+	if gt.spec.VersionFilter.Pattern != "" {
+		return false, files, message, fmt.Errorf("Target validation error: spec.versionfilter.pattern is not allowed for targets of type gittag.")
+	}
+
+	// Fail if the git tag resource cannot be validated
```
Comment (lemeurherve) [inline pkg/plugins/gittag/target.go:58]:
```suggestion
	// Fail if the git tag resource can't be validated
```

File: pkg/plugins/gittag/target.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+	err := gt.Validate()
+	if err != nil {
+		logrus.Errorln(err)
+		return false, files, message, err
+	}
+
+	// Check if the provided tag (from source input value) already exist
```
Comment (lemeurherve) [inline pkg/plugins/gittag/target.go:65]:
```suggestion
	// Check if the provided tag (from source input value) already exists
```

File: pkg/plugins/gittag/target.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	if gt.foundVersion.ParsedVersion == source {
+		// No error, but no change
+		logrus.Printf("%s The Git Tag %q already exists, nothing else to do.",
+			result.SUCCESS,
+			source)
+		return false, files, message, nil
+	}
+
+	// Otherwise proceed to creates this new tag
```
Comment (lemeurherve) [inline pkg/plugins/gittag/target.go:84]:
```suggestion
	// Otherwise proceed to create this new tag
```

File: pkg/plugins/gittag/target.go [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+	logrus.Printf("%s The Git Tag %q does not exist: creating it.", result.ATTENTION, source)
```
Comment (lemeurherve) [inline pkg/plugins/gittag/target.go:85]:
```suggestion
	logrus.Printf("%s The git tag %q does not exist: creating it.", result.ATTENTION, source)
```

File: pkg/plugins/gittag/target.go [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+	if dryRun {
+		// Dry run: no changes to apply.
+		// Return early without creating tag but notifies that a change should be made.
```
Comment (lemeurherve) [inline pkg/plugins/gittag/target.go:89]:
```suggestion
		// Return early without creating tag but notify that a change should be made.
```

File: pkg/plugins/gittag/target.go [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+		return true, files, message, nil
+	}
+
+	changed, err := generic.NewTag(source, gt.spec.Message, gt.spec.Path)
+	if err != nil {
+		return changed, files, message, err
+	}
+	logrus.Printf("%s The Git Tag %q was created.", result.ATTENTION, source)
```
Comment (lemeurherve) [inline pkg/plugins/gittag/target.go:97]:
```suggestion
	logrus.Printf("%s The git tag %q has been created.", result.ATTENTION, source)
```

**Review by lemeurherve — APPROVED** [review]
Small nipticks, LGTM otherwise!

File: pkg/plugins/gittag/main.go (hunk 7 above)
Comment (dduportal) [reply pkg/plugins/gittag/main.go:36]: After `did`, you must use the inifitive form ;)

Comment (dduportal) [reply pkg/plugins/gittag/main.go:36]: (Or we could say "Have you specified")

File: pkg/plugins/gittag/target.go (hunk 14 above)
Comment (dduportal) [reply pkg/plugins/gittag/target.go:58]: I'm not sure to see the reason to use a contracted form?

File: pkg/plugins/gittag/target.go (hunk 19 above)
Comment (dduportal) [reply pkg/plugins/gittag/target.go:97]: ack, even though I'm not sure if using the preterit form isn't best.

**Comment by dduportal** [comment]
Thanks for the review folks! I've enabled auto-merge (with squash).

I've taken in account almost all feedbacks (except 2: I've added comments).

As soon as you approve, this will be merged : subsequent comment will need ...a subsequent PR ;)

File: pkg/plugins/gittag/target.go (hunk 14 above)
Comment (lemeurherve) [reply pkg/plugins/gittag/target.go:58]: Less to read? 😛

**Review by lemeurherve — APPROVED** [review]

**Pull request merged**